test/test_tune.py

```diff
@@ -116,49 +116,102 @@ def test_autotune_match(self, open_db_mock: mock.Mock, close_db_mock: mock.Mock,
+        # Now test incremental fallback on device  [ driver version | platform | name ]
+        # when KATSDPSIGPROC_TUNE_MATCH is set to 'nearest'
         with mock.patch('katsdpsigproc.tune.KATSDPSIGPROC_TUNE_MATCH', 'nearest'):
-            # Re-run with different configurations, and make sure it always uses cached values.
+            # autotuning re-runs if arg_parameters are different
+            tuning_4 = {'a': 7, 'b': 8}
+            self.autotune_mock.return_value = tuning_4
+            context.device.driver_version = 'mock version'
+            context.device.platform_name = 'mock platform'
+            context.device.name = 'mock device'
+            ret = self.autotune(context, 'zzz')
+            assert ret == tuning_4
+
+            # Re-run with different configurations making sure the cached values are returned if
+            # they exist in any configuration, and that autotuning does not run.
             self.autotune_mock.side_effect = RuntimeError
 
             context.device.driver_version = 'abc'
             context.device.platform_name = 'mock platform'
             context.device.name = 'mock device'
-            ret4 = self.autotune(context, 'xyz')
-            assert ret4 == tuning
+            ret = self.autotune(context, 'xyz')
+            assert ret == tuning_1
 
             context.device.driver_version = 'abc'
             context.device.platform_name = 'abc'
             context.device.name = 'another device'
-            ret5 = self.autotune(context, 'xyz')
-            assert ret5 == another_tuning
+            ret = self.autotune(context, 'xyz')
+            assert ret == tuning_2
 
-            # If there is no match in the database, any tuning will suffice
-            context.device.driver_version = 'abc'
-            context.device.platform_name = 'abc'
-            context.device.name = 'abc'
-            ret6 = self.autotune(context, 'xyz')
-            assert ret6 == tuning or ret6 == another_tuning
+            context.device.driver_version = 'mock version'
+            context.device.platform_name = 'mock platform'
+            context.device.name = 'mock device'
+            ret = self.autotune(context, 'xyz')
+            assert ret == tuning_1
+
+
+class TestAutotunerNoParam:
+    """Tests for class that implements autotune with no parameters
+    """
+
+    autotune_mock = mock.Mock()
+
+    @pytest.fixture
+    def conn(self) -> Generator[sqlite3.Connection, None, None]:
+        conn = sqlite3.connect(':memory:')
+        yield conn
+        conn.close()
+
+    @pytest.fixture(autouse=True)
+    def _prepare_autotune_mock(self) -> Generator[None, None, None]:
+        # The autotune method has to be a class method due to magic in the
+        # autotuner decorator, and hence autotune_mock has to be at class
+        # scope. But we want it reset for each test.
+        self.autotune_mock.reset()
+        yield
+        self.autotune_mock.reset()
```

Comment:
There is a lot of duplicated code here. I don't think it's necessary to have a totally separate test class. Just have a second `autotuner`-decorated method (the method name `autotune` is not magic, so you can call it `autotune_no_args` or something).